Pandas Data Frames
==========

The DataFrame is the way Pandas represents a table, and Series is the data-structure Pandas use to represent a column. So, a data frame and series are synonomous with table and column.

What makes Pandas so attractive is the powerful interface to access individual records of the table, proper handling of missing values, and relational-databases operations between DataFrames.

In [1]:
# LOAD IN A DATA FRAME
from matplotlib import pyplot as plt
import pandas as pd
df = pd.read_csv('../data/gapminder_gdp_europe.csv', index_col='country')

## Inspecting Data


* We can print the first or last x number of rows of our data frame using the head() and tails() functions.

In [2]:
print(df.head(3))

         gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  gdpPercap_1967  \
country                                                                   
Albania     1601.056136     1942.284244     2312.888958     2760.196931   
Austria     6137.076492     8842.598030    10750.721110    12834.602400   
Belgium     8343.105127     9714.960623    10991.206760    13149.041190   

         gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
country                                                                   
Albania     3313.422188      3533.00391     3630.880722     3738.932735   
Austria    16661.625600     19749.42230    21597.083620    23687.826070   
Belgium    16672.143560     19117.97448    20979.845890    22525.563080   

         gdpPercap_1992  gdpPercap_1997  gdpPercap_2002  gdpPercap_2007  
country                                                                  
Albania     2497.437901     3193.054604     4604.211737     5937.029526  
Austria    27042.018680   

In [3]:
print(df.tail(3))

                gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
country                                                          
Switzerland       14734.232750    17909.489730    20431.092700   
Turkey             1969.100980     2218.754257     2322.869908   
United Kingdom     9979.508487    11283.177950    12477.177070   

                gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  \
country                                                          
Switzerland       22966.144320     27195.11304    26982.290520   
Turkey             2826.356387      3450.69638     4269.122326   
United Kingdom    14142.850890     15895.11641    17428.748460   

                gdpPercap_1982  gdpPercap_1987  gdpPercap_1992  \
country                                                          
Switzerland       28397.715120    30281.704590    31871.530300   
Turkey             4241.356344     5089.043686     5678.348271   
United Kingdom    18232.424520    21664.787670    22705.092540   

       

* We can use `dftypes` and `info()` to also get information about the data

In [4]:
print(df.dtypes)

gdpPercap_1952    float64
gdpPercap_1957    float64
gdpPercap_1962    float64
gdpPercap_1967    float64
gdpPercap_1972    float64
gdpPercap_1977    float64
gdpPercap_1982    float64
gdpPercap_1987    float64
gdpPercap_1992    float64
gdpPercap_1997    float64
gdpPercap_2002    float64
gdpPercap_2007    float64
dtype: object


In [5]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 30 entries, Albania to United Kingdom
Data columns (total 12 columns):
gdpPercap_1952    30 non-null float64
gdpPercap_1957    30 non-null float64
gdpPercap_1962    30 non-null float64
gdpPercap_1967    30 non-null float64
gdpPercap_1972    30 non-null float64
gdpPercap_1977    30 non-null float64
gdpPercap_1982    30 non-null float64
gdpPercap_1987    30 non-null float64
gdpPercap_1992    30 non-null float64
gdpPercap_1997    30 non-null float64
gdpPercap_2002    30 non-null float64
gdpPercap_2007    30 non-null float64
dtypes: float64(12)
memory usage: 3.0+ KB
None


* Use `describe()` to get basic statistical info about the data

In [6]:
print(df.describe())

       gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  gdpPercap_1967  \
count       30.000000       30.000000       30.000000       30.000000   
mean      5661.057435     6963.012816     8365.486814    10143.823757   
std       3114.060493     3677.950146     4199.193906     4724.983889   
min        973.533195     1353.989176     1709.683679     2172.352423   
25%       3241.132406     4394.874315     5373.536612     6657.939047   
50%       5142.469716     6066.721495     7515.733738     9366.067033   
75%       7236.794919     9597.220820    10931.085347    13277.182057   
max      14734.232750    17909.489730    20431.092700    22966.144320   

       gdpPercap_1972  gdpPercap_1977  gdpPercap_1982  gdpPercap_1987  \
count       30.000000       30.000000       30.000000       30.000000   
mean     12479.575246    14283.979110    15617.896551    17214.310727   
std       5509.691411     5874.464896     6453.234827     7482.957960   
min       2860.169750     3528.481305     3630.880

* Can also use describe() on data frame selections (like a single column)

In [164]:
print( df["gdpPercap_1982"].describe() )

count       30.000000
mean     15617.896551
std       6453.234827
min       3630.880722
25%      11449.870115
50%      15322.824720
75%      20901.729730
max      28397.715120
Name: gdpPercap_1982, dtype: float64


* Use `shape` to get the row and column numbers

In [165]:
print(df.shape)

(30, 12)


Here we can see the that the data have 30 rows of data and 12 attributes worth of information.

* Use the `len()` function to get numbers of each individually

In [166]:
# print number of rows of data
print(len(df))

30


In [167]:
# print number of columns of data
print(len(df.columns))

12


* Get the column names with `columns()`

In [168]:
print(df.columns)

Index(['gdpPercap_1952', 'gdpPercap_1957', 'gdpPercap_1962', 'gdpPercap_1967',
       'gdpPercap_1972', 'gdpPercap_1977', 'gdpPercap_1982', 'gdpPercap_1987',
       'gdpPercap_1992', 'gdpPercap_1997', 'gdpPercap_2002', 'gdpPercap_2007'],
      dtype='object')


* Use a column name to get all values for that columns

---
## Get information about a particular column

* Operations like mean, max, min, can be used on individual columns

In [56]:
# Mean GDP in 1967
print(df["gdpPercap_1967"].mean())
# Mean GDP in 1972
print(df["gdpPercap_1972"].mean())
# Mean GDP in 1977
print(df["gdpPercap_1977"].mean())


10143.823756533333
12479.57524646667
14283.9791096


## Rearange Columns

* Difficult to do using a csv library and doing this by hand
* The reverse() function will reverse the ordering of a list
     * E.g.   `['a', 'b', 'c']` to `['c', 'b', 'a']`

In [90]:
cols = list(df.columns)
print( cols )

cols.reverse()
print ( cols )

['gdpPercap_1952', 'gdpPercap_1957', 'gdpPercap_1962', 'gdpPercap_1967', 'gdpPercap_1972', 'gdpPercap_1977', 'gdpPercap_1982', 'gdpPercap_1987', 'gdpPercap_1992', 'gdpPercap_1997', 'gdpPercap_2002', 'gdpPercap_2007']
['gdpPercap_2007', 'gdpPercap_2002', 'gdpPercap_1997', 'gdpPercap_1992', 'gdpPercap_1987', 'gdpPercap_1982', 'gdpPercap_1977', 'gdpPercap_1972', 'gdpPercap_1967', 'gdpPercap_1962', 'gdpPercap_1957', 'gdpPercap_1952']


* Using that now reversed list above, we can create a new list, with the values order in reverse

In [169]:
new_df = df[cols]
new_df.head(3)

gdpPercap_2007  gdpPercap_2002  gdpPercap_1997  gdpPercap_1992  \
country                                                                   
Albania     5937.029526     4604.211737     3193.054604     2497.437901   
Austria    36126.492700    32417.607690    29095.920660    27042.018680   
Belgium    33692.605080    30485.883750    27561.196630    25575.570690   

         gdpPercap_1987  gdpPercap_1982  gdpPercap_1977  gdpPercap_1972  \
country                                                                   
Albania     3738.932735     3630.880722      3533.00391     3313.422188   
Austria    23687.826070    21597.083620     19749.42230    16661.625600   
Belgium    22525.563080    20979.845890     19117.97448    16672.143560   

         gdpPercap_1967  gdpPercap_1962  gdpPercap_1957  gdpPercap_1952  
country                                                                  
Albania     2760.196931     2312.888958     1942.284244     1601.056136  
Austria    12834.602400    10750.721110     8842.598030     6137.076492  
Belgium    13149.041190    10991.206760     9714.960623     8343.105127

## Selecting values

Remember that a DataFrame provides a index as a way to identify the rows of the table. A row also has a position inside the table as well as a label, which uniquely identifies its entry in the DataFrame.

To access a value at the position [ i , j ] (row, column) of a DataFrame, we have two options, depending on what is the meaning of i in use.



We can see that the first value in the first column is `1601.056136`. Remember that in programming, we begin our counting at zero, so this index would be `0,0`, not `1,1`.

### Use DataFrame.iloc[..., ...] to select values by their position
* Allows you to specify location by numerical index similar to 2D version of character selection in strings.


In [11]:
print("\nData value in first row at first column: ", df.iloc[0, 0])


Data value in first row at first column:  1601.056136


In [12]:
print("\nData value in fifth row at third column: ", df.iloc[4, 2])


Data value in fifth row at third column:  4254.337839


### Use `DataFrame.loc[..., ...]` to select values by their (entry) label.

*   Can specify location by name or by numerical index.

In [13]:
print(df.loc["Albania", "gdpPercap_1952"])

1601.056136


In [14]:
print(df.loc['Bulgaria', "gdpPercap_1962"])

4254.337839


### Use `:` on its own to mean all columns or all rows.

*   Just like Python's usual slicing notation, we can print all columns or all rows with `.loc` using the `:`

In [15]:
print(df.loc["Albania",:])

gdpPercap_1952    1601.056136
gdpPercap_1957    1942.284244
gdpPercap_1962    2312.888958
gdpPercap_1967    2760.196931
gdpPercap_1972    3313.422188
gdpPercap_1977    3533.003910
gdpPercap_1982    3630.880722
gdpPercap_1987    3738.932735
gdpPercap_1992    2497.437901
gdpPercap_1997    3193.054604
gdpPercap_2002    4604.211737
gdpPercap_2007    5937.029526
Name: Albania, dtype: float64


* Would get the same result printing `df.iloc[0]` (without a second index).
* We can also omit the `:` and get the same result in either case.
    * e.g. `df.loc["Albania"]`

In [16]:
print(df.loc[:, "gdpPercap_1952"])

country
Albania                    1601.056136
Austria                    6137.076492
Belgium                    8343.105127
Bosnia and Herzegovina      973.533195
Bulgaria                   2444.286648
Croatia                    3119.236520
Czech Republic             6876.140250
Denmark                    9692.385245
Finland                    6424.519071
France                     7029.809327
Germany                    7144.114393
Greece                     3530.690067
Hungary                    5263.673816
Iceland                    7267.688428
Ireland                    5210.280328
Italy                      4931.404155
Montenegro                 2647.585601
Netherlands                8941.571858
Norway                    10095.421720
Poland                     4029.329699
Portugal                   3068.319867
Romania                    3144.613186
Serbia                     3581.459448
Slovak Republic            5074.659104
Slovenia                   4215.041741
Spain            

*   Would get the same result printing `df["gdpPercap_1952"]`
*   Also get the same result printing `df.gdpPercap_1952` (since it's a column name)

### Select multiple columns or rows using `DataFrame.iloc` and a named slice.

In [17]:
print(df.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972'])

             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972
country                                                    
Italy           8243.582340    10022.401310    12269.273780
Montenegro      4649.593785     5907.850937     7778.414017
Netherlands    12790.849560    15363.251360    18794.745670
Norway         13450.401510    16361.876470    18965.055510
Poland          5338.752143     6557.152776     8006.506993


In the above code, we discover that **slicing using indexes is inclusive at both ends**, which differs from typical python behavior where slicing indicates everything up to but not including the final index.

However, if we use integers when our DataFrame is indexed by something else, slicing follows typical pythonic behavior.

## Result of slicing can be used in further operations.

In [18]:
print(df.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972'].max())

gdpPercap_1962    13450.40151
gdpPercap_1967    16361.87647
gdpPercap_1972    18965.05551
dtype: float64


In [19]:
print(df.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972'].min())

gdpPercap_1962    4649.593785
gdpPercap_1967    5907.850937
gdpPercap_1972    7778.414017
dtype: float64


*   Usually don't just print a slice.
*   All the statistical operators that work on entire data frames work the same way on slices.

## Create data frame from selections

We can create new data frame by selecting data frames based on values

In [150]:
# Use a subset of data to keep output readable.
subsetdf = df.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972']
print('Subset of data:\n', subset)

Subset of data:
              gdpPercap_1962  gdpPercap_1967  gdpPercap_1972
country                                                    
Italy           8243.582340    10022.401310    12269.273780
Montenegro      4649.593785     5907.850937     7778.414017
Netherlands    12790.849560    15363.251360    18794.745670
Norway         13450.401510    16361.876470    18965.055510
Poland          5338.752143     6557.152776     8006.506993


### Create DataFrame using query

* By passing a dataframe query to a itself, we can create a new dataframe with only those values

In [158]:
subset10kdf = subsetdf[subsetdf["gdpPercap_1962"] >= 10000]
# this is the same as subset[subset >= 10000]
#print("Type: ", type(subset10kdf))
print(subset10kdf)

             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972
country                                                    
Netherlands     12790.84956     15363.25136     18794.74567
Norway          13450.40151     16361.87647     18965.05551


*   Get the value where the mask is true, and NaN (Not a Number) where it is false.
*   Useful because NaNs are ignored by operations like max, min, average, etc.

## Filter a DataFrame using a Boolean mask

* A frame full of Booleans is sometimes called a *mask* because of how it can be used
* Comparison is applied element by element
* Returns a similarly-shaped data frame of `True` and `False`

In [160]:
mask10k = subsetdf >= 10000
print( mask10k )

             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972
country                                                    
Italy                 False            True            True
Montenegro            False           False           False
Netherlands            True            True            True
Norway                 True            True            True
Poland                False           False           False


* We can use masks to filter an entire dataframe with a single query
    * More efficient than using a single query on multiple columns

In [163]:
subset = subsetdf[mask10k]
print(subset)

             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972
country                                                    
Italy                   NaN     10022.40131     12269.27378
Montenegro              NaN             NaN             NaN
Netherlands     12790.84956     15363.25136     18794.74567
Norway          13450.40151     16361.87647     18965.05551
Poland                  NaN             NaN             NaN


## Quiz

* Create three data frames and get the size of each one.
    1. Countries with a gdp per capita above 10000 in 1952
    1. Countries with a gdp per capita above 10000 in 1962
    1. Countries with a gdp per capita above 10000 in 1972



In [176]:
df_1 = df[df["gdpPercap_1952"] > 10000]
df_1.shape

(2, 12)

In [177]:
df_1 = df[df["gdpPercap_1962"] > 10000]
df_1.shape

(12, 12)

In [178]:
df_1 = df[df["gdpPercap_1972"] > 10000]
df_1.shape

(19, 12)

## Create new columns

* We can easily create new columns in the same way we would add a key and value to a dictionary

In [185]:
# Create a new column diff_07_52 that is the difference between gdp per capita from 1952 to 2007
df["diff_07_52"] = df["gdpPercap_2007"] - df["gdpPercap_1952"]
df.head()

gdpPercap_1952  gdpPercap_1957  gdpPercap_1962  \
country                                                                  
Albania                    1601.056136     1942.284244     2312.888958   
Austria                    6137.076492     8842.598030    10750.721110   
Belgium                    8343.105127     9714.960623    10991.206760   
Bosnia and Herzegovina      973.533195     1353.989176     1709.683679   
Bulgaria                   2444.286648     3008.670727     4254.337839   

                        gdpPercap_1967  gdpPercap_1972  gdpPercap_1977  \
country                                                                  
Albania                    2760.196931     3313.422188     3533.003910   
Austria                   12834.602400    16661.625600    19749.422300   
Belgium                   13149.041190    16672.143560    19117.974480   
Bosnia and Herzegovina     2172.352423     2860.169750     3528.481305   
Bulgaria                   5577.002800     6597.494398     7612.240438   

                        gdpPercap_1982  gdpPercap_1987  gdpPercap_1992  \
country                                                                  
Albania                    3630.880722     3738.932735     2497.437901   
Austria                   21597.083620    23687.826070    27042.018680   
Belgium                   20979.845890    22525.563080    25575.570690   
Bosnia and Herzegovina     4126.613157     4314.114757     2546.781445   
Bulgaria                   8224.191647     8239.854824     6302.623438   

                        gdpPercap_1997  gdpPercap_2002  gdpPercap_2007  \
country                                                                  
Albania                    3193.054604     4604.211737     5937.029526   
Austria                   29095.920660    32417.607690    36126.492700   
Belgium                   27561.196630    30485.883750    33692.605080   
Bosnia and Herzegovina     4766.355904     6018.975239     7446.298803   
Bulgaria                   5970.388760     7696.777725    10680.792820   

                          diff_07_52  
country                               
Albania                  4335.973390  
Austria                 29989.416208  
Belgium                 25349.499953  
Bosnia and Herzegovina   6472.765608  
Bulgaria                 8236.506172

# Exercises

> ## Selection of Individual Values
>
> Assume Pandas has been imported into your notebook and the Gapminder GDP data for Europe has been loaded:
>
> ~~~
> import pandas
>
> df = pandas.read_csv('data/gapminder_gdp_europe.csv', index_col='country')
> ~~~
>
> Write an expression to find the Per Capita GDP of Serbia in 2007.

>> ### Solution:
>>The selection can be done by using the labels for both the row (“Serbia”) and the column (“gdpPercap_2007”):

>>`print(df.loc['Serbia', 'gdpPercap_2007'])`

>>The output is

>>`9786.534714`

> ## Extent of Slicing
>
> 1.  Do the two statements below produce the same output?
> 2.  Based on this,what rule governs what is included (or not) in numerical slices and named slices in Pandas?
>
> ~~~
> print(df.iloc[0:2, 0:2])
> print(df.loc['Albania':'Belgium', 'gdpPercap_1952':'gdpPercap_1962'])
> ~~~
> 
>>### Solution
>> No, they do not produce the same output! The output of the first statement is:

>>        gdpPercap_1952  gdpPercap_1957
>>country                                
>>Albania     1601.056136     1942.284244
>>Austria     6137.076492     8842.598030

>>The second statement gives:
~~~
>>           gdpPercap_1952  gdpPercap_1957  gdpPercap_1962 \
>>country                                                
>>Albania     1601.056136     1942.284244     2312.888958
>>Austria     6137.076492     8842.598030    10750.721110
>>Belgium     8343.105127     9714.960623    10991.206760
~~~
>>Clearly, the second statement produces an additional column compared to the first statement. What conclusion can we draw? We see that a numerical slice, 0:2, omits the final index (i.e. index 2) in the range provided, while a named slice, ‘gdpPercap_1952’:’gdpPercap_1962’, includes the final element

> ## Reconstructing Data
>
> Explain what each line in the following short program does:
> what is in `first`, `second`, etc.?
>
> ~~~
> first = pandas.read_csv('data/gapminder_gdp_all.csv', index_col='country')
> second = df[df['continent'] == 'Americas']
> third = second.drop('Puerto Rico')
> fourth = third.drop('continent', axis = 1)
> fourth.to_csv('result.csv')
> ~~~

>>### Solution
>>Let’s go through this piece of code line by line.
>>
>>`first = pandas.read_csv('data/gapminder_all.csv', index_col='country')`
>>
>>This line loads the dataset containing the GDP data from all countries into a dataframe called first. The index_col='country' parameter selects which column to use as the row labels in the dataframe.
>>
>>`second = first[first['continent'] == 'Americas']`
>>
>>This line makes a selection: only those rows of first for which the ‘continent’ column matches ‘Americas’ are extracted. Notice how the Boolean expression inside the brackets, first['continent'] == 'Americas', is used to select only those rows where the expression is true. Try printing this expression! Can you print also its individual True/False elements? (hint: first assign the expression to a variable)
>>
>>`third = second.drop('Puerto Rico')`
>>
>>As the syntax suggests, this line drops the row from second where the label is ‘Puerto Rico’. The resulting dataframe third has one row less than the original dataframe second.
>>
>>`fourth = third.drop('continent', axis = 1)`
>>
>>Again we apply the drop function, but in this case we are dropping not a row but a whole column. To accomplish this, we need to specify also the axis parameter (we want to drop the second column which has index 1).
>>
>>`fourth.to_csv('result.csv')`
>>
>>The final step is to write the data that we have been working on to a csv file. Pandas makes this easy with the to_csv() function. The only required argument to the function is the filename. Note that the file will be written in the directory from which you started the Jupyter or Python session.

> ## Selecting Indices
>
> Explain in simple terms what `idxmin` and `idxmax` do in the short program below.
> When would you use these methods?
>
> ~~~
> df = pandas.load_csv('data/gapminder_gdp_europe.csv', index_col='country')
> print(df.idxmin())
> print(df.idxmax())
> ~~~
>> ### Solution
>> For each column in data, idxmin will return the index value corresponding to each column’s minimum; idxmax will do accordingly the same for each column’s maximum value.

> ## Practice with Selection.
>
> Assume Pandas has been imported and the Gapminder GDP data for Europe has been loaded.
> Write an expression to select each of the following:
>
> 1.  GDP per capita for all countries in 1982.
> 2.  GDP per capita for Denmark for all years.
> 3.  GDP per capita for all countries for years *after* 1985.
> 4.  GDP per capita for each country in 2007 as a multiple of 
>     GDP per capita for that country in 1952.

>>### Solution
>>1: `df['gdpPercap_1982']`
>>
>>2: `df.loc['Denmark',:]`
>>
>>3: `df.loc[:,'gdpPercap_1985':]`
>>
>>4: `df['gdpPercap_2007']/data['gdpPercap_1952']`
>>

>> ## Interpretation
>>
>> Poland's borders have been stable since 1945, but changed several times in the years before then. How would you handle this if you were creating a table of GDP per capita for Poland for the entire Twentieth Century?

---
## Keypoints:
 - "Use `DataFrame.iloc[..., ...]` to select values by index location."
 - "Use `:` on its own to mean all columns or all rows."
 - "Select multiple columns or rows using `DataFrame.ix` and a named slice."
 - "Result of slicing can be used in further operations."
 - "Use comparisons to select data based on value."
 - "Select values or NaN using a Boolean mask."